From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
Date: Wed, 13 Jul 2022 20:31:49 +0300
Subject: [PATCH 14/17] [FIO extra] image-fit: dont use weak hash if
 SIGNATURE_STRICT

Disable weak methods of fit verification/signing when
SIGNATURE_STRICT enabled. This protects against potential threats
vectors.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
(cherry picked from commit a51b8b451975a22d61d47b1741f505c4f5a86c1c)
---
 boot/image-fit.c | 34 ++++++++++++++++++++++++++++++++++
 boot/image-sig.c |  2 +-
 2 files changed, 35 insertions(+), 1 deletion(-)

diff --git a/boot/image-fit.c b/boot/image-fit.c
index 3f1b7325c61..164b7c9be3d 100644
--- a/boot/image-fit.c
+++ b/boot/image-fit.c
@@ -1260,6 +1260,28 @@ int fit_set_timestamp(void *fit, int noffset, time_t timestamp)
 	return 0;
 }
 
+#if CONFIG_IS_ENABLED(FIT_SIGNATURE_STRICT)
+# define ALGO_NAME_ENT(x)	{ x, sizeof(x) }
+static struct algo_name {
+	const char *name;
+	int len;
+} loose_algo[] = {
+	ALGO_NAME_ENT("crc32"),
+	ALGO_NAME_ENT("sha1"),
+	ALGO_NAME_ENT("md5"),
+};
+
+int weak_algo(const char *name)
+{
+	for (int i = 0; i < sizeof(loose_algo) / sizeof(loose_algo[0]); i++) {
+		if (!strncmp(name, loose_algo[i].name, loose_algo[i].len))
+		    return -1;
+	}
+
+	return 0;
+}
+#endif
+
 /**
  * calculate_hash - calculate and return hash for provided input data
  * @data: pointer to the input data
@@ -1292,6 +1314,12 @@ int calculate_hash(const void *data, int data_len, const char *name,
 		return -1;
 	}
 
+#if CONFIG_IS_ENABLED(FIT_SIGNATURE_STRICT)
+	if (weak_algo(name)) {
+		debug("Only strong hash algorithm accepted\n");
+		return -1;
+	}
+#endif
 	hash_algo = hash_algo_lookup_by_name(name);
 	if (hash_algo == HASH_ALGO_INVALID) {
 		debug("Unsupported hash algorithm\n");
@@ -1309,6 +1337,12 @@ int calculate_hash(const void *data, int data_len, const char *name,
 	struct hash_algo *algo;
 	int ret;
 
+#if CONFIG_IS_ENABLED(FIT_SIGNATURE_STRICT)
+	if (weak_algo(name)) {
+		debug("Only strong hash algorithm accepted\n");
+		return -1;
+	}
+#endif
 	ret = hash_lookup_algo(name, &algo);
 	if (ret < 0) {
 		debug("Unsupported hash alogrithm\n");
diff --git a/boot/image-sig.c b/boot/image-sig.c
index 6bc74866eae..82e9b43cdfc 100644
--- a/boot/image-sig.c
+++ b/boot/image-sig.c
@@ -16,7 +16,7 @@ DECLARE_GLOBAL_DATA_PTR;
 #define IMAGE_MAX_HASHED_NODES		100
 
 struct checksum_algo checksum_algos[] = {
-#if CONFIG_IS_ENABLED(SHA1)
+#if CONFIG_IS_ENABLED(SHA1) && !CONFIG_IS_ENABLED(FIT_SIGNATURE_STRICT)
 	{
 		.name = "sha1",
 		.checksum_len = SHA1_SUM_LEN,
